BIT_+_ADR_UK_synth_data.ipynb

```diff
@@ -832,6 +832,9 @@
         "\n",
         "# Replace nulls/NaNs with string\n",
         "original_data.replace(np.nan, null_string, inplace = True)\n",
+        "for column in original_data.columns:\n",
+        "    if original_data[column].dtype != pd.Int64Dtype():\n",
```

Comment:
Does this mean that Int64 columns won't include the replacement strings?  I wonder if they should be converted to 'object' type at this stage so we can do that.

Reply:
With the original code I get errors like this:
```
Traceback (most recent call last):
  File "BIT_+_ADR_UK_synth_data.py", line 591, in <module>
    original_data.replace(np.nan, null_string, inplace = True)
  File "venv/lib/python3.10/site-packages/pandas/core/generic.py", line 8141, in replace
    new_data = self._mgr.replace(
  File "venv/lib/python3.10/site-packages/pandas/core/internals/base.py", line 249, in replace
    return self.apply_with_block(
  File "venv/lib/python3.10/site-packages/pandas/core/internals/managers.py", line 363, in apply
    applied = getattr(b, f)(**kwargs)
  File "venv/lib/python3.10/site-packages/pandas/core/internals/blocks.py", line 881, in replace
    putmask_inplace(blk.values, mask, value)
  File "venv/lib/python3.10/site-packages/pandas/core/array_algos/putmask.py", line 56, in putmask_inplace
    values[mask] = value
  File "venv/lib/python3.10/site-packages/pandas/core/arrays/masked.py", line 314, in __setitem__
    value = self._validate_setitem_value(value)
  File "venv/lib/python3.10/site-packages/pandas/core/arrays/masked.py", line 305, in _validate_setitem_value
    raise TypeError(f"Invalid value '{str(value)}' for dtype {self.dtype}")
TypeError: Invalid value '' for dtype Int64
```

I assumed the problem was that a NaN in an int column couldn't be replaced by a string because you can't put strings in an int column, but having checked it elsewhere pandas silently converts the whole column to `object` so we shouldn't see any problem. But we do. Any ideas?

Reply:
I've tried inplace=False, assigning to a different dataframe, and I've tried doing each column separately, but same exception.

The only thing I can think of now is to use `na_rep=null_string` in `to_csv()` but that might not help with `write_sav()` as per https://github.com/BehaviouralInsights/BIT-ADRUK-synthetic-data-tool/pull/2/commits/0044d6c6db845fa7549dc95c09b50d3c94be8547

Reply:
I've done some more tests using this code
```
import pandas as pd
import numpy as np
df = pd.DataFrame( {
    'int_pd':   [ 1,   pd.NA,   3, 4, 5],
    'int_np':   [ 1,   np.nan,  3, 4, 5],
    'float_pd': [ 1.0, pd.NA,   3, 4, 5],
    'float_np': [ 1.0, np.nan,  3, 4, 5],
    'str_pd':   [ '1', pd.NA,  '3', '4', '5'],
    'str_np':   [ '1', np.nan, '3', '4', '5'],
} )
print('-----\ndtypes before forcing int columns to Int64Dtype\n%s' % df.dtypes)
df['int_pd'] = df['int_pd'].astype(pd.Int64Dtype())
df['int_np'] = df['int_np'].astype(pd.Int64Dtype())
print('-----\ndtypes before replacing NaNs\n%s' % df.dtypes)
print('-----\nTable containing NaNs\n%s' % df)
#print(df.replace(np.nan, 'NNAA')) # TypeError: Invalid value 'NNAA' for dtype Int64
for col in df.columns:
    try:
        df[col] = df[col].astype(object).replace(np.nan, 'NANNY')
    except Exception as e:
        print('cannot replace in column %s (%s)' % (col,e))
print('-----\ndtypes after replacing NaNs\n%s' % df.dtypes)
df.to_csv('/dev/tty')
```

I've come to the conclusion that using `astype(object)` is the best solution here.
Oh, and using `inplace=True` gives warnings so assignment is used.
I've pushed a final commit https://github.com/BehaviouralInsights/BIT-ADRUK-synthetic-data-tool/pull/2/commits/e714a859801b07466d6fe440e9d6b4e73d8cc068 that I hope completes the changes.

Reply:
That looks great, thanks!